[ADD] sale_tiered_pricing: add tiered pricing

len-foss opened the pull request:
This PR adds support for tiered pricing.

~~(It depends on https://github.com/odoo/odoo/pull/60519)~~ already merged!

Info/Usage/... can be found in the description file.

Commit:
[ADD] sale_tiered_pricing: add tiered pricing

Commit:
[REM] sale_tiered_pricing: remove volume pricing option

This is already doable through standard pricelists.

Commit:
[IMP] sale_tiered_pricing: add tiered pricing wizard

Commit:
[IMP] sale_tiered_pricing: add (empty) README

Commit:
[TEST] sale_tiered_pricing: add a test on tier boundaries

lmignon commented on sale_tiered_pricing/models/sale_order_line.py:
```diff
@@ -0,0 +1,72 @@
+                self.price_unit, self.product_uom, product, self.product_uom_qty
+            )
+            computation = " + ".join("{} x {}".format(q, p) for q, p in qps)
+            tier_description = _("Tiers: {}").format(computation)
```
@len-foss `_()` will not be translated into the partner lang..... How can we force the use of the partner lang when calling the translate method?

len-foss replied:
Damn, I thought getting the product in the right context would do that since it's what's done in standard.
I'll write a test case and check that, thanks for bringing that up.

rousseldenis replied:
Ok, this case is tested

lmignon reviewed (approved):
LGTM (Code + functional)  A nice to have would be to be able to get access to the tiered definitions from the product....

Commit:
[TEST] sale_tiered_pricing: test wizard, small improvements

ccasula commented:
Test case 1: Failed: SO (Ordered quantity = 0, description field displays 1)
![TC001](https://user-images.githubusercontent.com/12275185/98967928-5709e180-250d-11eb-8200-a8deaa436143.png)

ccasula commented:
Test case 2: Failed: SO: Rounding issue in the calculated price (it should be 1.157 instead of 1.156,95).

![TC002](https://user-images.githubusercontent.com/12275185/98968500-0a72d600-250e-11eb-8e1c-04885038da76.png)

Commit:
[IMP] sale_tiered_pricing: direct access on tiered pricing items from products

Note that we remove the exclusion on "pricelist on product" and
"pricelist with formulas" by making both implied by the root pricelist
group.
This makes the view on product accessible, with access to formulas.

len-foss commented:
Regarding the rounding issue: this is a known problem with this approach (I'd have to check in details the decimal precisions etc to see if there is a way to mitigate this). 
But I believe the trade-off regarding the simplicity that if confers with editing sale orders (no need to go through a wizard) as well as the code complexity if we had to split the lines makes it up for the issue.

For the description, the case "zero units" is degenerate and may add an extra if, but the overall multiline_description_sale is still TODO (a bit too spartan for the moment).

Commit:
[IMP] sale_tiered_pricing: translate tier computation in partner language

TODO: write test.

Commit:
[FIX] sale_tiered_pricing: fix test rounding

The rounding issues can be avoided by setting a high decimal accuracy
on the product price.
Note that an ORM bug makes this brittle in test environment
(it cannot be set into the test setup, however it works inside the test).

Commit:
[FIX] sale_tiered_pricing: no pricelist on sale order

Commit:
[FIX] sale_tiered_pricing: no tier computation on line with zero quantity

Commit:
[IMP] sale_tiered_pricing: filter partner pricelist field

Commit:
[FIX] sale_tiered_pricing: correct description in composite tiered pricings

Commit:
[IMP] sale_tiered_pricing: add pre-init-hook to speed up install

Commit:
[IMP] sale_tiered_pricing: add demo data

Commit:
[IMP] sale_tiered_pricing: add context to skip tiered description

Commit:
[FIX] sale_tiered_pricing: update sale line description when quantity is updated

rousseldenis commented on sale_tiered_pricing/wizard/tiered_pricing_wizard.py:
```diff
@@ -0,0 +1,108 @@
+
+    def create_pricelist_items(self):
+        if not self.pricelist_ids:
+            raise OdooWarning(_("You should select at least one target pricelist!"))
```
@len-foss Warning is deprecated. You should use UserError

rousseldenis reviewed (changes requested):
This seems great!

Little change
